File: src/main/python/Cleaner.py
import sys

test_case = """
3
3 2
0 -1
-1 2
-1 5
3 3
0 -1 -1
-1 2 -1
-1 -1 8
4 3
0 1 2
-1 -1 3
-1 -1 -1
-1 -1 11


"""


def parse_line():
    return [int(num) for num in sys.stdin.readline().split()]


def recursive_visit(m, t, r, depth, target, s):
    if t < 0 or r < 0 or len(m) <= t or len(m[0]) <= r:
        return 0
    if m[t][r] != -1 and m[t][r] != depth:
        return 0
    if depth == target and t == len(m) - 1 and r == len(m[0]) - 1:
        return 1

    origin_cell_value = m[t][r]
    m[t][r] = depth

    path_count = 0
    path_count += recursive_visit(m, t + 1, r, depth + 1, target, s)
    path_count += recursive_visit(m, t - 1, r, depth + 1, target, s)
    path_count += recursive_visit(m, t, r + 1, depth + 1, target, s)
    path_count += recursive_visit(m, t, r - 1, depth + 1, target, s)

    s += path_count
    s %= 100000007

    m[t][r] = origin_cell_value

    return s


cases = parse_line()[0]
for i in range(cases):
    x, y = parse_line()
    room_map = [[0] * y] * x
    for j in range(x):
        room_map[j] = parse_line()

    print(recursive_visit(room_map, 0, 0, 0, len(room_map) * len(room_map[0]) - 1, 0))

File: src/main/java/Cleaner.java
import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.io.InputStreamReader;

public class Cleaner {

    public static void main(String[] args) throws Exception {
        InputStream source = new ByteArrayInputStream(TestCases.CLEANER.getBytes());

        BufferedReader bi = new BufferedReader(new InputStreamReader(source));
        int cc = Integer.parseInt(bi.readLine().split(" ")[0]);
        for (int i = 0; i < cc; i++) {
            String[] s = bi.readLine().split(" ");
            int[][] map = new int[Integer.parseInt(s[0])][Integer.parseInt(s[1])];

            for (int j = 0; j < map.length; j++) {
                String[] ss = bi.readLine().split(" ");
                for (int k = 0; k < ss.length; k++) {
                    map[j][k] = Integer.parseInt(ss[k]);
                }
            }

            System.out.println(solution(map));
        }
    }

    public static long solution(int[][] map) {
        return recursiveVisit(map,
                0, 0, 0,
                map.length * map[0].length - 1,
                0);
    }

    public static long recursiveVisit(int[][] map,
                                      int x, int y,
                                      int depth, int target,
                                      long sum) {
        if (x < 0 || y < 0 || map.length <= x || map[0].length <= y) return 0;
        if (map[x][y] != -1 && map[x][y] != depth) return 0;
        if (depth == target && x == map.length - 1 && y == map[0].length - 1) return 1;

        int originCell = map[x][y];
        map[x][y] = depth;

        int answer = 0;
        answer += recursiveVisit(map,x + 1, y, depth + 1, target, sum);
        answer += recursiveVisit(map,x - 1, y, depth + 1, target, sum);
        answer += recursiveVisit(map, x, y + 1, depth + 1, target, sum);
        answer += recursiveVisit(map, x, y - 1, depth + 1, target, sum);

        sum += answer;
        sum %= 1000000007;

        map[x][y] = originCell;

        return sum;
    }
}
